### PR TITLE
fix(jsonschema): allow embed resources

### PR DESCRIPTION
**Pull request by soyuka**
#5793 related ping @paullallier @vincentchalamon

### CODE REVIEW

**Comment by paullallier** [comment]
That looks a lots better! A couple things I still see:

There's something odd about the reference back from TestEntity --> BagOfTests in my example. I don't ever user this in my API, so it might be normal - not sure.
Screenshot 2023-09-12 at 10 44 44

Also, I haven't defined the return structure on errors.  In alpha2, it didn't try to display them but in this branch it is displaying as something undefined.  This might be a feature of the updated SwaggerUI, and it might not be a bad things, since I probably should defined the error return... Though perhaps I shouldn't need to do this explicitly?

In my dev environment, I have the API configured to display on https://main-api (it also displays on https://main-api/docs).  In alpha2, it works.  In this branch, I get raw json (LD, I think) on https://main-api (the /docs version give the expect html).

**Comment by paullallier** [comment]
Oh, and I think that was with e03bb31

File: src/JsonSchema/SchemaFactory.php [hunk 1]
```diff
@@ -194,8 +201,18 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
...
-            $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
-            // prevent "type" and "anyOf" conflict
+            if ($isCollection) {
+                $propertySchema['items']['$ref'] = $subSchema['$ref'];
+                unset($propertySchema['items']['type']);
+                break;
+            }
+
+            if ($type->isNullable()) {
+                $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
```
Comment (vincentchalamon) [inline src/JsonSchema/SchemaFactory.php:211]: IMHO `oneOf` should be more persistent in this case: https://json-schema.org/understanding-json-schema/reference/combining.html#id7
```suggestion
                $propertySchema['oneOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
```

File: src/JsonSchema/SchemaFactory.php [hunk 2]
```diff
@@ -194,8 +201,18 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
...
+            } else {
+                $propertySchema['oneOf'] = [['$ref' => $subSchema['$ref']]];
```
Comment (vincentchalamon) [inline src/JsonSchema/SchemaFactory.php:213]: Useless `oneOf` in this case, `$ref` is enough (cf. https://json-schema.org/understanding-json-schema/structuring.html#ref)

File: src/JsonSchema/SchemaFactory.php [hunk 3]
```diff
@@ -176,13 +182,14 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
 
         // property schema is created in SchemaPropertyMetadataFactory, but it cannot build resource reference ($ref)
         // complete property schema with resource reference ($ref) only if it's related to an object
-
         $version = $schema->getVersion();
         $subSchema = new Schema($version);
         $subSchema->setDefinitions($schema->getDefinitions()); // Populate definitions of the main schema
+        $isCollection = false;
```
Comment (vincentchalamon) [inline src/JsonSchema/SchemaFactory.php:188]: Useless variable as it's only used inside the foreach loop

File: src/JsonSchema/SchemaFactory.php [hunk 4]
```diff
@@ -194,8 +201,18 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
             }
 
             $subSchema = $this->buildSchema($className, $format, Schema::TYPE_OUTPUT, null, $subSchema, $serializerContext + [self::FORCE_SUBSCHEMA => true], false);
-            $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
-            // prevent "type" and "anyOf" conflict
+            if ($isCollection) {
```
Comment (vincentchalamon) [inline src/JsonSchema/SchemaFactory.php:204]: Useless `$isCollection` var as it's only used inside this foreach loop.
Also, if the first foreach loop enables the `$isCollection` variable but the same loop exists on line 200, the next `$type` will be considered as a collection even if it isn't.
```suggestion
            if ($type->isCollection()) {
```

**Review by vincentchalamon — CHANGES_REQUESTED** [review]

File: src/JsonSchema/SchemaFactory.php (hunk 1 above)
Comment (soyuka) [reply src/JsonSchema/SchemaFactory.php:211]: you can probably have null values in a collection with objects?

Comment (vincentchalamon) [reply src/JsonSchema/SchemaFactory.php:211]: Like `['/books/{id}, null]`? Not sure this is a valid value for us, and is what we want...

**Pull request merged**